pandaserver/dataservice/finisher.py

```diff
@@ -0,0 +1,284 @@
+                data=json_data,
+            )
+
+    def check_file_status(self, job):
+        """
+        This function checks the status of the files for the job.
+
+        Parameters:
+        job (JobSpec): The job specification.
+
+        Returns:
+        tuple: A tuple containing three elements:
+            - bool: True if all files are ready, False otherwise.
+            - list: A list of failed files.
+            - list: A list of files with no output.
+        """
+        tmp_log = LogWrapper(_logger,
+                             f"check_file_status-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+        failed_files = []
+        no_out_files = []
+        for file in job.Files:
+            if file.type in ('output', 'log'):
+                if file.status == "failed":
+                    failed_files.append(file.lfn)
+                elif file.status == "nooutput":
+                    no_out_files.append(file.lfn)
+                elif file.status != "ready":
+                    tmp_log.debug(f"Job: {job.PandaID} file:{file.lfn} {file.status} != ready")
+                    return False, failed_files, no_out_files
+        return True, failed_files, no_out_files
+
+    def get_bit_map(self, required_tokens: str, dest_token: str, tmp_source_site_spec,
+                    tmp_dst_site_spec):
+        """
+        This function calculates the bit_map, comp_bit_map, and updated_bit_map based on the
+        required tokens, destination token, source site spec and destination site spec.
+
+        Parameters:
+        required_tokens (str): The required tokens.
+        dest_token (str): The destination token.
+        tmp_source_site_spec (SiteSpec): The source site specification.
+        tmp_dst_site_spec (SiteSpec): The destination site specification.
+
+        Returns:
+        tuple: A tuple containing the calculated bit_map and comp_bit_map.
+        """
+        # make bit_map for the token
+        bit_map = 1
+        if len(required_tokens.split(",")) > 1:
+            for tmp_req_token in required_tokens.split(","):
+                if tmp_req_token == dest_token:
+                    break
+                # shift one bit
+                bit_map <<= 1
+        # completed bit_map
+        comp_bit_map = (1 << len(required_tokens.split(","))) - 1
+        # ignore the lowest bit for T1, file on DISK is already there
+        if tmp_source_site_spec.ddm_output == tmp_dst_site_spec.ddm_output:
+            comp_bit_map = comp_bit_map & 0xFFFE
+        return bit_map, comp_bit_map
+
+    # main
+    def run(self):
+        """
+        Starts the thread to finish transferring jobs
+        """
+        tmp_log = LogWrapper(_logger,
+                             f"run-{datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None).isoformat('/')}")
+        # start
+        try:
+            by_call_back = False
+            if self.job is None:
```

Comment:
I'm still trying to figure out what the bitmap function does. So I wanted to see the print out and concluded that no ATLAS panda server went into this "if" condition since the ALMA9 migration. We should discuss with @tmaeno quickly after the Tuesday meeting if this part is obsolete and if not, try to understand the bit map logic to document it better.

Reply:
The bitmap stuff was required to place output files both to DISK and TAPE at the same time. The mechanism has not been used for >10 years, so you can remove that part.

Reply:
Just to be sure lines 188 to 247 can be removed?
if self.job is None:
...
            else:

If confirmed and to be doublechecked, but that would also obsolete:
`updateTransferStatus` and `updateOutFilesReturnPandaIDs` in Taskbuffer and OraDBProxy

Reply:
Thank you Tadashi, I will remove it but I can see the following on updateTransferStatus (OraDBProxy):
            sqlU = 'UPDATE /*+ INDEX_RS_ASC(TAB("DATASETS"."NAME")) */ ATLAS_PANDA.Datasets tab SET transferStatus=ATLAS_PANDA.BITOR(transferStatus,:bitMap) WHERE name=:name'

I guess the call to updateTransferSTatus is still required, right? Anyway, I would much rather prefer to go through the required updated logic together if you both have some time after Tuesday's core meeting.

Reply:
> Just to be sure lines 188 to 247 can be removed? if self.job is None: ... else:
> 
> If confirmed and to be doublechecked, but that would also obsolete: `updateTransferStatus` and `updateOutFilesReturnPandaIDs` in Taskbuffer and OraDBProxy

Ok got it. If confirmed I will remove these calls (and delete them from TaskBuffer and OraDBProxy.